crates/noirc_evaluator/src/ssa/opt/array_use.rs

```diff
@@ -0,0 +1,51 @@
+use std::collections::HashMap;
+
+use crate::ssa::{
+    ir::{
+        basic_block::BasicBlockId,
+        dfg::DataFlowGraph,
+        instruction::{Instruction, InstructionId},
+        post_order::PostOrder,
+        value::ValueId,
+    },
+    ssa_gen::Ssa,
+};
+
+impl Ssa {
+    /// Map arrays with the last instruction that uses it
+    /// For this we simply process all the instructions in execution order
+    /// and update the map whenever there is a match
+    pub(crate) fn array_use(&self) -> HashMap<ValueId, InstructionId> {
+        let mut array_use = HashMap::new();
+        for func in self.functions.values() {
+            let mut reverse_post_order = Vec::new();
+            reverse_post_order.extend_from_slice(PostOrder::with_function(func).as_slice());
+            reverse_post_order.reverse();
```

Comment:
We can avoid cloning here if we add an `into_vec` method for `PostOrder`

Reply:
Done